lib/src/client/http2_connection.dart

```diff
@@ -96,17 +111,45 @@ class Http2ClientConnection implements connection.ClientConnection {
       return;
     }
     _setState(ConnectionState.connecting);
-    connectTransport().then((transport) {
+    connectTransport().then((transport) async {
       _currentReconnectDelay = null;
       _transportConnection = transport;
+      _connectionLifeTimer
+        ..reset()
+        ..start();
       transport.onActiveStateChanged = _handleActiveStateChanged;
       _setState(ConnectionState.ready);
-      _pendingCalls.forEach(_startCall);
-      _pendingCalls.clear();
+
+      if (_hasPendingCalls()) {
+        // Take all pending calls out, and reschedule.
+        final pendingCalls = _pendingCalls.toList();
+        _pendingCalls.clear();
+        pendingCalls.forEach(dispatchCall);
+      }
+
     }).catchError(_handleConnectionFailure);
   }
 
+  /// Abandons the current connection if it is unhealthy or has been open for
+  /// too long.
+  ///
+  /// Assumes [_transportConnection] is not `null`.
+  void _refreshConnectionIfUnhealthy() {
+    final bool isHealthy = _transportConnection.isOpen;
+    final bool shouldRefresh =
+        _connectionLifeTimer.elapsed > options.connectionTimeout;
+    if (!isHealthy || shouldRefresh) {
+      if (isHealthy) {
```

Comment:
This is the same as `if (shouldRefresh)`, and there's no reason for it to be in the outer `if` block.